pkg/envoy/lds/listener.go

```diff
@@ -57,38 +53,41 @@ func (lb *listenerBuilder) newOutboundListener() (*xds_listener.Listener, error)
 		} else if egressTrafficPolicy != nil {
 			egressFilterChains := lb.getEgressFilterChainsForMatches(egressTrafficPolicy.TrafficMatches)
 			listener.FilterChains = append(listener.FilterChains, egressFilterChains...)
-			filterDisableMatchPredicate = getFilterMatchPredicateForTrafficMatches(egressTrafficPolicy.TrafficMatches)
+			trafficMatches = append(trafficMatches, egressTrafficPolicy.TrafficMatches...)
 		}
-		additionalListenerFilters := []*xds_listener.ListenerFilter{
-			// Configure match predicate for ports serving server-first protocols (ex. mySQL, postgreSQL etc.).
-			// Ports corresponding to server-first protocols, where the server initiates the first byte of a connection, will
-			// cause the HttpInspector ListenerFilter to timeout because it waits for data from the client to inspect the protocol.
-			// Such ports will set the protocol to 'tcp-server-first' in an Egress policy.
-			// The 'FilterDisabled' field configures the match predicate.
-			{
-				// To inspect TLS metadata, such as the transport protocol and SNI
-				Name:           wellknown.TlsInspector,
-				FilterDisabled: filterDisableMatchPredicate,
-			},
-			{
-				// To inspect if the application protocol is HTTP based
-				Name:           wellknown.HttpInspector,
-				FilterDisabled: filterDisableMatchPredicate,
-			},
-		}
-		listener.ListenerFilters = append(listener.ListenerFilters, additionalListenerFilters...)
-
-		// ListenerFilter can timeout for server-first protocols. In such cases, continue the processing of the connection
-		// and fallback to the default filter chain.
-		listener.ContinueOnListenerFiltersTimeout = true
+	}
+	trafficMatches = append(trafficMatches, lb.meshCatalog.GetOutboundMeshTrafficPolicy(lb.serviceIdentity).TrafficMatches...)
+	filterDisableMatchPredicate = getFilterMatchPredicateForTrafficMatches(trafficMatches)
+	listener.ListenerFilters = []*xds_listener.ListenerFilter{
+		{
+			// The OriginalDestination ListenerFilter is used to redirect traffic
+			// to its original destination.
+			Name: wellknown.OriginalDestination,
+		},
+		// Configure match predicate for ports serving server-first protocols (ex. mySQL, postgreSQL etc.).
+		// Ports corresponding to server-first protocols, where the server initiates the first byte of a connection, will
+		// cause the HttpInspector ListenerFilter to timeout because it waits for data from the client to inspect the protocol.
+		// Such ports will set the protocol to 'tcp-server-first' in an Egress policy.
+		// The 'FilterDisabled' field configures the match predicate.
+		{
+			// To inspect TLS metadata, such as the transport protocol and SNI
+			Name:           wellknown.TlsInspector,
+			FilterDisabled: filterDisableMatchPredicate,
+		},
+		{
+			// To inspect if the application protocol is HTTP based
+			Name:           wellknown.HttpInspector,
+			FilterDisabled: filterDisableMatchPredicate,
+		},
```

Comment:
These 2 filters are only required when EgressPolicy is enabled. Why is that behavior being changed?

Reply:
Didn't try to change anything extra like that. I'll change it back.

Reply:
It's fine to have them but it would result in a behavior change, where these 2 filters could time out for non-egress traffic, which was previously not the case.